src/Workspaces/CoreTest/Execution/Extensions.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.UnitTests.Execution
+{
+    internal static class Extensions
+    {
+        public static ChecksumObjects<ProjectChecksumObject> ToProjectObjects(this ChecksumCollection collection, ISolutionChecksumService service)
+        {
+            Contract.ThrowIfFalse(collection.Kind == WellKnownChecksumObjects.Projects);
+            return new ChecksumObjects<ProjectChecksumObject>(service, collection);
+        }
+
+        public static ChecksumObjects<DocumentChecksumObject> ToDocumentObjects(this ChecksumCollection collection, ISolutionChecksumService service)
+        {
+            Contract.ThrowIfFalse(collection.Kind == WellKnownChecksumObjects.Documents || collection.Kind == WellKnownChecksumObjects.TextDocuments);
+            return new ChecksumObjects<DocumentChecksumObject>(service, collection);
+        }
+    }
+
+    internal class ChecksumObjects<T> : ChecksumObject where T : ChecksumObject
```

Comment:
>ChecksumObjects [](start = 19, length = 15)

This is confusing between ChecksumObjects and ChecksumCollection. As far as I can tell, ChecksumCollection == a collection of Checksums and ChecksumObjects == a collection of ChecksumObjects. Why do we need both instead of just one or the other? Actually, why do we need this since it doesn't allow WriteToAsync?